### PR TITLE
WIP: Rethink CLI

### LINKED ISSUE #5: Figure out optimal CLI commands
OK, the way I see it, we have 4 major commands (with `profile` having 2 main options):
```sh
$ ff download [-c {channel}]
$ ff install [-c {channel}]
$ ff uninstall [-c {channel}]
$ ff profile --create {name}
$ ff profile --delete {name}
```

I'm not 100% sold on how `ff profile -c/--create` should work though. Currently when creating a profile, we need the following bits of information:
```py
def create_mozprofile(application, test_type, env, profile_dir):
```

Where, the args are as follows:
- `application`: name of an application, ie: "loop-server"
- `test_type`: directory of the test type we want to run, ie: "stack-check"
- `env`: testing environment we plan on using, ie: "dev", "stage", or "prod" (or something else)
- `profile_dir`: name of the directory to save the profile to (which will have a prefs.js and user.js -- or something close to that).

We have those first 3 args, so we can do something like load the following config files:
- "prefs.ini"
- "{application}/prefs.ini:{env}"
- "{application}/{test_type}/prefs.ini:{env}"

I don't know how we want to pass those 4 args.

Not sure if we need to add one more nested depth of commands for `ff profile` so it's something like:
```sh
$ ff profile create --application loop-server --test_type stack-check --env stage --name rpapa9000
```

otherwise, we may be able to do something like this:
```sh
$ ff profile --create rpapa9000 --application loop-server --test_type stack-check --env stage
# ...
$ ff profile --delete rpapa9000
```

Not sure if that is awkward since we're treating `--create` and `--delete` as sub-subparsers, but binding the profile name/directory to it.

If we go with the latter option of `ff profile --delete {name}`, not sure how awkward it may be if `ff profile --help` gives you a bunch of non-relevant CLI options for `--delete` when we really only want the name of the profile to delete.

DISCUSS!

### PR DESCRIPTION
**Pull request by pdehaan**
FIxes #5

### CODE REVIEW

**Comment by pdehaan** [comment]
Let's use https://docs.python.org/3.5/library/tempfile.html#tempfile.mkdtemp to manage creating a temporary profile directory in _temp/profiles if a profile name was not supplied. Or something.

**Comment by pdehaan** [comment]
We can launch w/ specific profile using the `-profile` flag and a fully qualified path to the profile dir, like a so:

```sh
$ ./_temp/browsers/FirefoxNightly.app/Contents/MacOS/firefox-bin -profile "./_temp/profiles/fftool.1dDKz8"
```

File: configs/darwin.ini [hunk 1]
```diff
@@ -1,24 +1,28 @@
...
-PATH_FIREFOX_APP = ""
+APP_SRC_FILENAME = Firefox.app
+APP_DEST_FILENAME = FirefoxBeta.app
 
 [release]
 DOWNLOAD_FILENAME = FirefoxRelease.dmg
-APP_FILENAME = Firefox.app
-PATH_FIREFOX_APP = ""
+APP_SRC_FILENAME = Firefox.app
+APP_DEST_FILENAME = FirefoxRelease.app
```
Comment (pdehaan) [inline configs/darwin.ini]: We currently only have the OSX/Darwin config in here, but we'll need to make similar tweakage to the Windows/Cygwin and Linux flavors.

File: fftool/firefox_download.py [hunk 2]
```diff
@@ -69,17 +68,12 @@ def download(channel):
     is_recent_file = modification_date(download_path) > SCRIPT_START_TIME
 
     if is_recent_file:
+        print("You should install")
         install(channel)
+    else:
+        print("Not recent. skipping install")
```
Comment (pdehaan) [inline fftool/firefox_download.py]: nit: remove the debug nonsense.

File: fftool/firefox_download.py [hunk 3]
```diff
@@ -69,17 +68,12 @@ def download(channel):
     is_recent_file = modification_date(download_path) > SCRIPT_START_TIME
```
Comment (rpappalax) [inline fftool/firefox_download.py]: doesn't mozdownloaden check this for us?  not sure if is_recent_file was something legacy we had from before

File: fftool/firefox_profile.py [hunk 4]
```diff
@@ -10,58 +10,66 @@
...
+                    valid_paths.append(path_app_test_type + ":" + env)
 
+    return valid_paths
 
-def create_mozprofile(application, test_type, env, profile_dir):
 
-    full_profile_dir = os.path.join(PATH_TEMP, profile_dir)
+def create_mozprofile(profile_dir, application=None, test_type=None, env=None):
+    if not profile_dir:
+        print("Create random profile")
```
Comment (pdehaan) [inline fftool/firefox_profile.py]: nit: remove debug `print()` statements from L57 and L59 and L61

File: fftool/firefox_profile.py [hunk 5]
```diff
@@ -10,58 +10,66 @@
...
+        full_profile_dir = mkdtemp(dir=PROFILE_DIR, prefix="fftool.", suffix="")
+        print(full_profile_dir)
+    else:
+        print("Using named profile")
+        full_profile_dir = os.path.join(PROFILE_DIR, profile_dir)
 
     # If temp profile already exists, kill it so it doesn't merge unexpectedly.
```
Comment (rpappalax) [inline fftool/firefox_profile.py]: I believe this comment is obsolete

File: fftool/main.py [hunk 6]
```diff
@@ -12,39 +14,83 @@
...
+    if not options.no_launch:
+        print("Launching!")
+        launch_firefox(profile_path, channel=options.channel)
+
+    """
+    if not options.env:
+        print("Unknown env")
+    else:
+        print("Load settings for env: {0}".format(options.env))
+    """
```
Comment (pdehaan) [inline fftool/main.py]: Remove the commented out block above, and possibly all the unnecessary `print()` statements above.

File: fftool/firefox_run.py [hunk 7]
```diff
@@ -14,12 +15,24 @@
...
+
+
+def launch_firefox(profile_path, channel):
+    """
+    this will be an all-encompassing function that may make use of all the
+    others (with the exception of uninstall
+    """
 
-    # we need to invoke each of these here 
+    # we need to invoke each of these here
```
Comment (rpappalax) [inline fftool/firefox_run.py]: we should probably distribute these comments into their respective modules so we're printing them out as they actually happen, but i'm ok w/ leaving her from now

File: requirements.txt [hunk 8]
```diff
@@ -2,5 +2,7 @@ Fabric3
 configargparse
 configparser
 mozdownload==1.20.2
+mozprofile
 tox==2.3.1
 flake8==2.5.4
+outlawg
```
Comment (pdehaan) [inline requirements.txt]: nit: Pin to specific versions?
Also, do we need both `configargparse` and `configparser`? Seems redundant redundant.

File: fftool/main.py [hunk 9]
```diff
@@ -12,39 +14,83 @@
...
+    print("Creating profile...")
+
+    profile_path = create_mozprofile(options.profile, options.app, options.test_type, options.env)
+
+    # LAUNCH
+    if not options.no_launch:
+        print("Launching!")
+        launch_firefox(profile_path, channel=options.channel)
+
+    """
```
Comment (rpappalax) [inline fftool/main.py]: we can drop these lines methinks

**Comment by rpappalax** [comment]
apart from some tweaky nits, nice work man!

**Comment by pdehaan** [comment]
Please update the README.md

File: fftool/firefox_download.py [hunk 10]
```diff
@@ -63,23 +64,25 @@ def download(channel):
...
+    print("Downloading {channel} to {download_path}".format(**args))
     scraper.download()
 
     is_recent_file = modification_date(download_path) > SCRIPT_START_TIME
+    firefox_bin = env.get(channel, 'PATH_FIREFOX_BIN_ENV')
 
-    if is_recent_file:
+    # If the *.dmg file was downloaded recently, or we don't have the *.app
+    # file installed, install the current Firefox channel.
+    if is_recent_file or not os.path.exists(firefox_bin):
```
Comment (pdehaan) [inline fftool/firefox_download.py]: @rpappalax This was one of the tweaks I made which checks if the *.dmg was downloaded in this "session" OR if the *.app is not found in the specified directory, THEN we mount the *.dmg and install the *.app. This should solve that edge case we were seeing where the *.dmg was the latest, so the *.app wasn't getting installed always.

**Comment by rpappalax** [comment]
:+1:

**Pull request merged**